shell/platform/linux/fl_view.cc

```diff
@@ -153,6 +154,60 @@ static gboolean run_flutter_engine(FlView* self) {
   return TRUE;
 }
 
+static gboolean fl_view_send_pointer_button_event(FlView* self,
+                                                  GdkEventButton* event) {
+  FlutterPointerEvent fl_event = {};
+  fl_event.struct_size = sizeof(fl_event);
+
+  int64_t button;
+  switch (event->button) {
+    case 1:
+      button = kFlutterPointerButtonMousePrimary;
+      break;
+    case 2:
+      button = kFlutterPointerButtonMouseMiddle;
+      break;
+    case 3:
+      button = kFlutterPointerButtonMouseSecondary;
+      break;
+    default:
+      return FALSE;
+  }
+  int old_button_state = self->button_state;
+  if (event->type == GDK_BUTTON_PRESS) {
+    // Drop the event if Flutter already thinks the button is down
+    if ((self->button_state & button) != 0)
+      return FALSE;
+    self->button_state ^= button;
+
+    if (old_button_state == 0)
+      fl_event.phase = kDown;
+    else
+      fl_event.phase = kMove;
+  } else if (event->type == GDK_BUTTON_RELEASE) {
+    // Drop the event if Flutter already thinks the button is up
+    if ((self->button_state & button) == 0)
+      return FALSE;
+    self->button_state ^= button;
+
+    if (self->button_state == 0)
+      fl_event.phase = kUp;
+    else
+      fl_event.phase = kMove;
+  }
+
+  if (self->flutter_engine == nullptr)
+    return FALSE;
+
+  fl_event.x = event->x;
+  fl_event.y = event->y;
+  fl_event.device_kind = kFlutterPointerDeviceKindMouse;
+  fl_event.buttons = self->button_state;
```

Comment:
I just noticed you're not sending the timestamp; please add that so the event is complete.

Reply:
How sensitive is Flutter to the timestamp? I was reading some code that suggested it should be from the same clock as Flutter is using, but then the other implementations don't seem to be very careful about that. I have a timestamp from GTK that I'm hoping is suitable, but not 100% sure. I imagine in practise all these timestamps are ultimately sourced from the kernel monotonic timer, so in practise, any timestamp will work.

Reply:
The uses I'm aware of are things like calculating fling velocity, so I would expect we'll be fine with the GTK timestamps. Let's use those for now, and revisit in the future if we ever have a concrete example of needing something else.

Reply:
GDK provides timestamps in milliseconds, and I convert these to microseconds as per the coment in `shell/platform/embedder/embedder.h`

```
  /// The timestamp at which the pointer event was generated. The timestamp
  /// should be specified in microseconds and the clock should be the same as
  /// that used by `FlutterEngineGetCurrentTime`.
  size_t timestamp;
```

Interestingly the macos code seems to set them to nanoseconds:

```
.timestamp = static_cast<size_t>(event.timestamp * NSEC_PER_MSEC)
```

So not sure if the code or the comment is wrong.

Reply:
Entertainingly, `event.timestamp` is in seconds, so that turns out to be right even though it makes no sense. Thanks for noticing that; I've sent you a PR to fix it. (My guess is an off-by-one in copying a constant from a header maybe.)